File: backend/handlers/auth_handler.go
package handlers

import (
	"net/http"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/golang-jwt/jwt/v5"
	"github.com/jabatkopi/backend/database"
	"github.com/jabatkopi/backend/models"
	"golang.org/x/crypto/bcrypt"
)

type LoginRequest struct {
	Username string `json:"username" binding:"required"`
	Password string `json:"password" binding:"required"`
}

var jwtSecret = []byte("super-secret-key-for-jabatkopi")

func LoginHandler(c *gin.Context) {
	var req LoginRequest

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"status":  400,
			"message": "Format data tidak valid. Silakan coba lagi.",
		})
		return
	}

	var user models.User
	if err := database.DB.Where("username = ?", req.Username).First(&user).Error; err != nil {
		c.JSON(http.StatusUnauthorized, gin.H{
			"status":  401,
			"message": "Maaf, username atau password salah. Silakan periksa kembali.",
		})
		return
	}

	if err := bcrypt.CompareHashAndPassword([]byte(user.PasswordHash), []byte(req.Password)); err != nil {
		c.JSON(http.StatusUnauthorized, gin.H{
			"status":  401,
			"message": "Maaf, username atau password salah. Silakan periksa kembali.",
		})
		return
	}

	token := jwt.NewWithClaims(jwt.SigningMethodHS256, jwt.MapClaims{
		"user_id":  user.ID,
		"username": user.Username,
		"role":     user.Role,
		"exp":      time.Now().Add(time.Hour * 72).Unix(),
	})

	tokenString, err := token.SignedString(jwtSecret)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"status":  500,
			"message": "Gagal membuat sesi masuk. Silakan coba lagi.",
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"status":  200,
		"message": "Berhasil masuk!",
		"data": gin.H{
			"token": tokenString,
			"role":  user.Role,
		},
	})
}

func ChangePasswordHandler(c *gin.Context) {
	var req struct {
		OldPassword string `json:"old_password" binding:"required"`
		NewPassword string `json:"new_password" binding:"required,min=6"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"message": "Password minimal 6 karakter."})
		return
	}

	userID, _ := c.Get("user_id")
	var user models.User
	if err := database.DB.First(&user, userID).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"message": "User tidak ditemukan."})
		return
	}

	if err := bcrypt.CompareHashAndPassword([]byte(user.PasswordHash), []byte(req.OldPassword)); err != nil {
		c.JSON(http.StatusUnauthorized, gin.H{"message": "Password lama Anda tidak sesuai."})
		return
	}

	hashedPassword, _ := bcrypt.GenerateFromPassword([]byte(req.NewPassword), bcrypt.DefaultCost)
	user.PasswordHash = string(hashedPassword)
	database.DB.Save(&user)

	c.JSON(http.StatusOK, gin.H{"message": "Password berhasil diperbarui!"})
}

func RegisterHandler(c *gin.Context) {
	var req struct {
		Name     string `json:"name" binding:"required"`
		Username string `json:"username" binding:"required"`
		Email    string `json:"email" binding:"required,email"`
		Password string `json:"password" binding:"required,min=6"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"message": "Data tidak lengkap atau format salah (Password min. 6 karakter)."})
		return
	}

	var existingUser models.User
	if err := database.DB.Where("username = ? OR email = ?", req.Username, req.Email).First(&existingUser).Error; err == nil {
		c.JSON(http.StatusConflict, gin.H{"message": "Username atau Email sudah terdaftar. Gunakan yang lain."})
		return
	}

	hashedPassword, _ := bcrypt.GenerateFromPassword([]byte(req.Password), bcrypt.DefaultCost)

	user := models.User{
		Name:         req.Name,
		Username:     req.Username,
		Email:        req.Email,
		PasswordHash: string(hashedPassword),
		Role:         "customer",
	}

	if err := database.DB.Create(&user).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"message": "Terjadi gangguan pada server. Silakan coba lagi nanti."})
		return
	}

	c.JSON(http.StatusCreated, gin.H{
		"status":  201,
		"message": "Pendaftaran berhasil! Silakan masuk.",
	})
}

File: lib/core/network/handlers/auth_handler.dart
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:supabase_flutter/supabase_flutter.dart';
import 'package:bcrypt/bcrypt.dart';

class AuthHandler {
  final FlutterSecureStorage secureStorage;

  AuthHandler(this.secureStorage);

  Future<dynamic> handle(
      String method, String cleanPath, Map<String, dynamic>? body) async {
    if (cleanPath == '/auth/login' && method == 'POST' && body != null) {
      final email = body['email'] as String;
      final password = body['password'] as String;

      final response = await Supabase.instance.client
          .from('users')
          .select()
          .eq('email', email)
          .maybeSingle();

      if (response == null) {
        throw Exception('Maaf, email tidak ditemukan.');
      }

      final passwordHash = response['password_hash'] as String?;
      if (passwordHash == null || !BCrypt.checkpw(password, passwordHash)) {
        throw Exception(
            'Maaf, email atau password salah. Silakan periksa kembali.');
      }

      final token =
          'session_${response['id']}_${DateTime.now().millisecondsSinceEpoch}';
      return {
        'status': 200,
        'message': 'Berhasil masuk!',
        'data': {
          'id': response['id'],
          'token': token,
          'role': response['role'] ?? 'customer',
          'username': response['username'] ?? 'User',
        }
      };
    }

    if (cleanPath == '/auth/register' && method == 'POST' && body != null) {
      final name = body['name'] as String;
      final username = body['username'] as String;
      final email = body['email'] as String;
      final password = body['password'] as String;

      final existing = await Supabase.instance.client
          .from('users')
          .select()
          .or('username.eq.$username,email.eq.$email')
          .maybeSingle();

      if (existing != null) {
        throw Exception(
            'Username atau Email sudah terdaftar. Gunakan yang lain.');
      }

      final passwordHash = BCrypt.hashpw(password, BCrypt.gensalt());
      await Supabase.instance.client.from('users').insert({
        'name': name,
        'username': username,
        'email': email,
        'password_hash': passwordHash,
        'role': 'customer',
        'created_at': DateTime.now().toIso8601String(),
        'updated_at': DateTime.now().toIso8601String(),
      });

      return {
        'status': 201,
        'message': 'Pendaftaran berhasil! Silakan masuk.',
      };
    }

    if (cleanPath == '/profile' && method == 'GET') {
      final currentUserIdStr = await secureStorage.read(key: 'user_id');
      final customerId = int.tryParse(currentUserIdStr ?? '') ?? 0;

      final user = await Supabase.instance.client
          .from('users')
          .select()
          .eq('id', customerId)
          .single();

      return {
        'status': 200,
        'data': {
          'id': user['id'],
          'name': user['name'] ?? '',
          'username': user['username'] ?? '',
          'email': user['email'] ?? '',
          'image_url': user['image_url'] ?? '',
          'role': user['role'] ?? 'customer',
        }
      };
    }

    if (cleanPath == '/profile' && method == 'PUT' && body != null) {
      final currentUserIdStr = await secureStorage.read(key: 'user_id');
      final customerId = int.tryParse(currentUserIdStr ?? '') ?? 0;

      await Supabase.instance.client
          .from('users')
          .update({
            'name': body['name'],
            'username': body['username'],
            'email': body['email'],
            'image_url': body['image_url'],
            'updated_at': DateTime.now().toIso8601String(),
          })
          .eq('id', customerId);

      return {
        'status': 200,
        'message': 'Profil berhasil diperbarui!',
      };
    }

    if (cleanPath == '/profile/password' && method == 'PUT' && body != null) {
      final oldPassword = body['old_password'] as String;
      final newPassword = body['new_password'] as String;

      final currentUserIdStr = await secureStorage.read(key: 'user_id');
      final customerId = int.tryParse(currentUserIdStr ?? '') ?? 0;

      final user = await Supabase.instance.client
          .from('users')
          .select()
          .eq('id', customerId)
          .single();
      final passwordHash = user['password_hash'] as String?;

      if (passwordHash == null || !BCrypt.checkpw(oldPassword, passwordHash)) {
        throw Exception('Password lama Anda tidak sesuai.');
      }

      final newPasswordHash = BCrypt.hashpw(newPassword, BCrypt.gensalt());
      await Supabase.instance.client
          .from('users')
          .update({
            'password_hash': newPasswordHash,
            'updated_at': DateTime.now().toIso8601String(),
          })
          .eq('id', customerId);

      return {
        'status': 200,
        'message': 'Password berhasil diperbarui!',
      };
    }

    throw Exception('Route not handled in AuthHandler: $method $cleanPath');
  }
}
